### ISSUE TITLE
Create performance benchmark for Deferreds

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3704]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#3704|
|Type|enhancement|
|Created|2009-03-25 20:03:53Z|
|Branch|https://github.com/twisted/twisted/tree/defer-benchmarks-3704|

James Knight has previously [sandbox/foom/cdefer implemented Deferred in Pyrex](sandbox/foom/cdefer implemented Deferred in Pyrex). The reason this isn't included in Twisted is that we just have no reliable, accurate way to measure what the performance changes of this are. The same applies to cdefer.c in [ticket:2245 Deferred implementation in C].

It would useful to have a benchmark which makes it possible to determine the utility of a Pyrex or C implementation of Deferred. Without actually measuring, statements about how fast Pyrex, C or pure Python is are worthless.

Deriving something from the code which showed a 5% - 20% speedup might be a good way to start on such a benchmark. It's important to cover all uses of Deferreds here. It's also okay if some parts show a slowdown, as long as the overall improvement is enough. Of course, it's better if everything is faster. :)

(ticket description derived from exarkun's comment in [#2245](https://github.com/twisted/twisted/issues/2245))

Attachments:

* [defer.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c1a/c1af44fdd5fad14eba26292e64a9c65d6dc25613/4494c814c06140612bde4d9aa53d4ee9cb29ddc6.py) (4000 bytes) - added by thijs on 2009-03-25 20:05:23Z - defer benchmark from source:branches/cdefer-2245-4/doc/core/benchmarks (#2245)
* [timer.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/c1a/c1af44fdd5fad14eba26292e64a9c65d6dc25613/a94f44d4d5a2520805de2ab59eb15221001bd13b.py) (440 bytes) - added by thijs on 2009-03-25 20:08:32Z - timer benchmark utility from source:branches/cdefer-2245-4/doc/core/benchmarks (#2245)

<details><summary>Searchable metadata</summary>

```
trac-id__3704 3704
type__enhancement enhancement
reporter__thijs thijs
priority__normal normal
milestone__ 
branch__branches_defer_benchmarks_3704 branches/defer-benchmarks-3704
branch_author__exarkun__peaker__thijs exarkun, peaker, thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__benchmark benchmark
time__1238011433000000 1238011433000000
changetime__1238683200000000 1238683200000000
version__None None
owner__ 
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra removed owner|
|-|-|

I attached [[defer.py]](None/defer.py]) and [[timer.py]](None/timer.py]) from the latest [branches/cdefer-2245-4/doc/core/benchmarks/ branch](branches/cdefer-2245-4/doc/core/benchmarks/ branch) for ticket [#2245](https://github.com/twisted/twisted/issues/2245), that run a ```twisted.internet.defer``` benchmark, with outputting:

```
$ python defer.py Disabled GC
twisted.internet.defer
instantiate () 0.205390930176
instantiateShootCallback () 0.780241012573
instantiateShootErrback () 0.100492954254
instantiateAddCallbacksNoResult (10,) 0.00270199775696
instantiateAddCallbacksNoResult (1000,) 0.263583898544
instantiateAddCallbacksNoResult (10000,) 3.02256608009
instantiateAddCallbacksBeforeResult (10,) 0.00721502304077
instantiateAddCallbacksBeforeResult (1000,) 0.712515115738
instantiateAddCallbacksBeforeResult (10000,) 13.2036309242
instantiateAddCallbacksAfterResult (10,) 0.00922584533691
instantiateAddCallbacksAfterResult (1000,) 0.903044939041
instantiateAddCallbacksAfterResult (10000,) 8.97325801849
pauseUnpause (10,) 0.00899577140808
pauseUnpause (1000,) 0.912666082382
pauseUnpause (10000,) 15.3666989803
```

I suggest adding defer.py (and updating timer.py in the [doc/core/benchmarks trunk](doc/core/benchmarks trunk)).

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_2"></a>|@moshez commented|
|-|-|

Adding branch field

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Plans for a benchmark api are described in [#3715](https://github.com/twisted/twisted/issues/3715)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

I think that branch field setting is erroneous.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26577]) Branching to 'defer-benchmarks-3704'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_6"></a>|@thijstriemstra commented|
|-|-|

(In [26594]) Add copyright header refs [#3704](https://github.com/twisted/twisted/issues/3704)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @exarkun|
|-|-|

I made some trivial tweaks.  I think this is good enough to merge.  I'm sure we'll want to continue changing things as we get more benchmarking experience.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [26630]) Merge defer-benchmarks-3704 

Author: Peaker, thijs, exarkun
Reviewer: exarkun
Fixes: [#3704](https://github.com/twisted/twisted/issues/3704)

Add a benchmark for Deferred performance.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
